<div id="helpSectionWrapper">

<h2>P-WI Variant Editor Interface functionality</h2>

<h4>General Description</h4>
<p>
The Variant module is a search, edit and display page all in one. Each variant is attached to a single allele, while an
allele can have multiple variants. Variants don't have an accession number, so if two alleles have the exact same variant, two variant records will have to be created (one for each allele). Details for multiple variants of one allele can be viewed and edited one at a time only.

</p><h4>Keyboard Short-cuts</h4>
There are reported issues using these short-cuts with Mac/Firefox. 
<ul>
	<li>Search = Ctrl-Alt-S</li>
	<li>Clear = Ctrl-Alt-C</li>
	<li>Modify = Ctrl-Alt-M</li>
	<li>Previous = Ctrl-Alt-P</li>
	<li>Next = Ctrl-Alt-N</li>
</ul>

<h4>Color-coded Labels</h4>
<ul>
        <li>Thistle = search only</li>
        <li>Green = search &amp; edit</li>
        <li>Blue = read-only</li>
        <li>Yellow = current row</li>
</ul>

<h4>Supported Browsers</h4>
<ul>
	<li>Chrome and FireFox are supported on both PC and Mac computers.</li>
	<li>Safari is NOT supported for the PWI. There are features that do not work.</li>
</ul>

<h4>Searching</h4>
<p>
All fields that have labels highlighted in <span style="background-color: #D8BFD8">
Thistle</span> may be used for searching. They may be searched individually or in 
combination. At least one search field must be entered or an error message is displayed. When more than one field is used in a search, it 
is treated as an AND search. (i.e.  <u>all</u> entered fields must match the variants to be returned.) 
</p>

<h5>Limit on # of results</h5>
<ul>
	<li>The Variant module is currently configured to restrict the number of results to 1,001 variants. It may be preferable to use more search fields to reduce the number of results returned.</li>
</ul> 

<h5>Wildcards</h5>
<ul>
	<li>Wildcards (%) may be used for most fields, except for <i>Creation Date</i> and <i>Modification Date</i>.</li>
	<li>For <i>Allele Symbol</i> a wildcard may be used at the beginning and/or end of your search text. E.g. %&lt;b2b% or %Btlr&gt; or Adgre1%. </li><li>For <i>Allele ID</i> and <i>Allele References</i>, a wildcard can only be used at the end, e.g. MGI:57895% or J:1752%. </li>

</ul> 

<h5>Search by Accession IDs</h5>
<ul>
	<li>Valid types of IDs: Allele ID with MGI: prefix or Allele Reference ID with J: prefix.</li>
	<li>For <i>Allele ID</i> only single IDs are accepted. For <i>Allele Reference</i> IDs, lists are accepted (space, comma, newline/carriage return, or tab separated).</li>
	
	
</ul>

<h5>Search by Dates</h5>
<ul>
	<li>The Creation and Modification Dates are time stamps held in the database.</li>
	
	<li>Valid date formats for searching are:
		<ul>
			<li>mm/dd/yyyy or mm-dd-yyyy</li>
			<li>mm/dd/yy or mm-dd-yy</li>
			<li>yyyy/mm/dd or yyyy-mm-dd</li>
		</ul>
	</li>
	<li>Searching for a range of dates:
		<ul>
			<li>equal to: enter a single date</li>
			<li>greater than: &gt; mm/dd/yyyy</li>
			<li>greater than or equal to: &gt;= mm/dd/yyyy</li>
			<li>less than: &lt; mm/dd/yyyy</li>
			<li>less than or equal to: &lt;= mm/dd/yyyy</li>
			<li>between two dates (inclusive): mm/dd/yyyy..mm/dd/yyyy</li>
			</ul>
	</li>
</ul> 

<h4>Search Results</h4>
<p>When you click on the Search button, the results are displayed in a table with each row representing a single
	variant.  The table displays the variants, through the alleles they are attached to, in alpha-numeric order. The variant highlighted in pale yellow (by default the first one) is the one that is
	displayed in detail in the main form. You can highlight a different variant by clicking on a different row.
	</p>
		
<h4>Adding New Data</h4>
<ul>
    <li>The <i>Reference Allele</i> and <i>Variant Allele</i> fields are used to enter/display the relevant sequence. 
        <ul><li>In the <i>genomic</i> and <i>transcript</i> sections, these fields accept only ATCGN+- for now. Wobble codes will follow. Note that + and - can only used on their own. Any spaces will be stripped out and lower case will be changed to upper case.</li><li>In the <i>polypeptide</i> section, these fields only accept the single amino-acid codes shown in the <b><font color="blue">Amino Acid Popup</font></b> window. Any spaces will be stripped out and lower case will be changed to upper case.</li></ul>
</li><li>To create a new variant from scratch, start with a blank form (click <b>Clear</b> if necessary) and enter either
an <i>Allele ID</i> or <i>Allele Symbol</i> in their respective fields. Then either hit the TAB key or use the mouse to 
click out of the field to any other field. The missing allele value (either <i>ID</i> or <i>Symbol</i>) plus associated 
data (<i>References</i>, <i>Chromosome</i>, <i>Strand</i>) will now be retrieved and filled in. Allele data can now be filled 
in (see Sourced and Curated Records Requirements below for requirements). A variant record will only be created after clicking 
<b>Create</b>. Subsequent changes are saved by clicking <b>Modify</b> (the <b>Create</b> button will be grayed out once the record has been created).</li>
<li>To create a variant that is a copy of (or very similar to) an existing variant, but has to be attached to a different allele of the same gene, 
click the <b>Duplicate Variant</b> button on the form of the variant record you want to duplicate. This will create a new form with all the variant data of 
the original form but with only the allele gene symbol filled in for the allele data. Complete the <i>Allele Symbol</i> to represent the new allele you want 
the variant to be attached to. Then either hit the TAB key or use the mouse to click out of the field to any other field. The missing allele value (<i>Allele ID</i>) 
plus associated data (<i>References</i>, <i>Chromosome</i>, <i>Strand</i>) will now be retrieved and filled in. Modify variant data as necessary. A variant record will only be 
created after clicking <b>Create</b>. Subsequent changes are saved by clicking <b>Modify</b> (the <b>Create</b> button will be grayed out once the record has been created).</li>
<li>To create a new variant for an existing allele (i.e. you have an allele variant record open), click the <b>Duplicate Allele</b> button. A new form will be created with just 
the allele data from the previous form filled in; the variant fields will be empty. Allele data can now be filled in (see Sourced and Curated Records Requirements below 
for requirements). A variant record will only be created after clicking <b>Create</b>. Subsequent changes are saved by clicking <b>Modify</b> (the <b>Create</b> button will be grayed out once 
the record has been created). Once the new record has been created the All variants table at the bottom will be updated with the new row of variant data added to the end of the table.</li>
<li>If you only provide Sourced info, then both Sourced and Curated records will be created but the Curated record will be empty; if you provide both Sourced and Curated info, then both 
records will be created and populated using Sourced and Curated info.
<ul><li>This means Sourced and Curated records will always be create regardless of whether you provide only Sourced or only Curated data, or both.</li></ul></li> 
</ul>
<h5>Modifying fields</h5>
<ul>
<li>All fields with a green label tab (except <i>Types</i> and <i>Effects</i>) accept keyboard or copy-pasted input.</li>
<li>To copy values from the As Sourced fields to the corresponding Curated fields, you can click the <b>Copy all</b> button to copy all data or click the <b>Coords</b> or <b>Alleles</b> button in the Genomic, 
Transcript or Polypeptide sections to copy only selected data. The <b>Coords</b> button copies only the coordinate data (<i>Genome Build/Transcript ID/Polypeptide ID, Start, End</i>) in that section to 
the corresponding Curated section; The <b>Alleles</b> button copies only the Reference and Variant Allele data (sequence) in that section to the corresponding Curated section.</li>
<li>To enter amino acid sequence you can click the <b><font color="blue">Amino Acid Popup</font></b> link. This will open a popup window with all the valid amino acids and their three-letter and one-letter codes. You can use this window to 
enter amino acids or use it just as a reference when using manual keyboard entry. If the former, click on any name or code to copy that value to the field that is active (i.e. with a pale yellow background). 
Note that the system will interpret full names and three-letter codes as strings of single-letter amino acid codes and, upon saving, capitalize all characters. So make sure that in the Polypeptide <i>Reference</i>
and <i>Variant Allele</i> fields you only use single-letter codes (upper or lower case). Full names or three-letter codes can be used in either of the <i>Notes</i> fields. </li>
</ul>

<h4>Allele Section</h4>
<ul>
<li>All fields are searchable.</li>
<li>A wildcard is allowed in the <i>Allele Symbol</i> field for searching.</li>
<li><i>Allele References</i> are displayed as a comma-separated list.</li>
<li>The <i>Strand</i> field will have a red background when the strand is "-".</li>

</ul>
<h4>Sourced and Curated Records Section Requirements</h4>
<h5>Genomic</h5>
<ul>
<li><i>Genome Build</i> must exist if you have <i>Start</i> and <i>Stop</i> coordinates.</li>
<li>Both  <i>Start</i> and <i>Stop</i> coordinates must be provided.</li><li>Genome <i>Start</i> coordinate must be less than or equal to the <i>Stop</i> coordinate.</li>

    <li>In the <i>Allele</i> fields you can enter nucleotide sequence in either upper or lower case and with or without spaces; lower case letters are automatically capitalized and spaces stripped out upon saving.</li>
</ul>
<h5>Transcript</h5>
<ul>
<li>If a transcript sequence has a <i>Start</i> and <i>Stop</i> coordinate, it must have an <i>ID</i> (which will be validated).</li>
<li>Valid <i>Transcript ID</i>s are RefSeq, Ensembl, or Genbank sequences. Use only unversioned IDs (i.e. omit the dot-version-number after the primary ID).</li><li>Both <i>Start</i> and <i>Stop</i> coordinates must be provided.</li>
<li>Transcript <i>Start</i> coordinate must be less than or equal to the <i>Stop</i> coordinate.</li>
<li>In the <i>Allele</i> fields you can enter nucleotide sequence in either upper or lower case and with or without spaces; lower case letters are automatically capitalized and spaces stripped out upon saving.</li>
</ul>
<h5>Polypeptide</h5>
<ul>
<li>If a polypeptide sequence has a <i>Start</i> and <i>Stop</i> coordinate, it must have an <i>ID</i> (which will be validated).</li>

<li>Both <i>Start</i> and <i>Stop</i> coordinates must be provided.</li><li>Polypeptide <i>Start</i> coordinate bust be less than or equal to the <i>Stop</i> coordinate.</li>
<li>In the <i>Allele</i> fields you can enter amino acid sequence in either upper or lower case and with or without spaces; lower case letters are automatically capitalized and spaces stripped out upon saving.</li>
</ul>
<h4>HGVS Notation</h4>
<ul>
<li>Displays the Human Genome Variation Society notation for the variant.</li>
</ul>
<h4>Variant References</h4>
<ul>
<li>The reference(s) from which the variant is sourced. This field can have one or more references.</li>
<li>The reference J numbers are validated.</li>
</ul>
<h4>Public and Private Notes</h4>
<ul>
<li><i>Public Notes</i> are used to communicate information about the variant to public users.</li>
<li><i>Curator Notes</i> are private and used to document information obtained from the curation process.</li>
</ul>
<h4>Variant Types and Effects</h4>
<ul>
<li>Click the <b><font color="blue">Types Popup</font></b> link to bring up a popup window with a selection of variant types. Click the check boxes in front of the types you want to apply to the variant and click <b>Save Changes and Close</b> to save the selection. The selected types will be filled in in the <i>Types</i> field.  Alternatively, click <b>Discard Changes and Close</b> (or the window close button top right) if you do not want to save the selection. If there is existing data in the <i>Types</i> field, those SO terms will show up as selected in the popup window. Any modifications to existing selections will be reflected in the field after save and close. A variant can be of more than one type. Type SO terms are shown with the most common first, logically grouped.</li>
<li>Click the <b><font color="blue">Effects Popup</font></b> link to bring up a popup window with a selection of variant effects. Click the check boxes in front of the effects you want to apply to the variant and click <b>Save Changes and Close</b> to save the selection. The selected effects will be filled in in the <i>Effects</i> field. Alternatively, click <b>Discard Changes and Close</b> (or the window close button top right) if you do not want to save the selection. If there is existing data in the <i>Effects</i> field, those SO terms will show up as selected in the popup window. Any modifications to existing selections will be reflected in the field after save and close. A variant can have more than one effect. Effect SO terms are shown with the most common first, logically grouped, followed by the less common terms sorted alphabetically.</li>
</ul>
<h4>All Variants Table</h4>
<ul>
<li>Displays a summary of variant data for all annotated variants attached to a given allele. </li>
<li><b><font color="blue">Allele symbol</font></b> above the table links to Allele detail page in MGI.</li>
<li>Genomic information contains Genome build, start coordinate, end coordinate, reference allele and variant allele, if provided.</li>
<li>Transcript and Polypeptide information contains Accession ID, start coordinate, end coordinate, reference allele and variant allele, if provided.</li>
<li>Curated data will be displayed if available, if no Curated data is available, Sourced data will be displayed.</li>
<li>Color-coded boundaries around the fields distinguish between Curated (<span style="background-color: #00FF00">green</span>) and Sourced (<span style="background-color: #FFFF00">yellow</span>).</li>
<li>Table is sorted by genome start, then transcript start.</li>
<li>Active row (by default the first one) will be highlighted with a pale yellow background; the full data for that variant is displayed in the main form.</li>
<li>The Variants table is read only (no editing).</li>
<li><b><font color="blue">Transcript IDs</font></b> and <b><font color="blue">Polypeptide IDs</font></b> are hyperlinks to their relevant sites.</li>
</ul>
</body></html>

</div> <!-- helpSectionWrapper -->
