<div id="helpSectionWrapper">

<h2>P-WI Allele Relationship Editor Interface functionality</h2>

<h4>General Description</h4>
<p>
The Allele Relationship module is a search, display and edit page all in one. A user must be logged in to edit but can search without logging in.
</p>
<h4>Keyboard Short-cuts</h4>
There are reported issues using these short-cuts with Mac/Firefox. 
<ul>
	<li>Search = Ctrl-Alt-S</li>
	<li>Clear = Ctrl-Alt-C</li>
	<li>Modify = Ctrl-Alt-M</li>
	<li>Previous = Ctrl-Alt-P</li>
	<li>Next = Ctrl-Alt-N</li>
</ul>

<h4>Color-coded Labels</h4>
<ul>
        <li>Thistle = search only</li>
        <li>Green = search &amp; edit</li>
        <li>Blue = read-only</li>
        <li>Yellow = current row</li>
        <li>Orange = Marker Region Tool</li>
</ul>

<h4>Supported Browsers</h4>
<ul>
	<li>Chrome and FireFox are supported on both PC and Mac computers.</li>
	<li>Safari is NOT supported for the PWI. There are features that do not work.</li>
</ul>
<h3>Pulldown Menus</h3>
<p>
Relationship type: descendants of Mutation_Involves (RV:0000003) or Expresses_Component (RV:0000004)<br>
Chr: mouse chromosome<br>
Organism Pick List: all Organisms in the MGI relationship set<br>
Property: Organism, Gene Symbol, and accepted IDs for expressed orthologous genes
</p>
<h3>Searching</h3>
<ul>
<li>All fields that have labels highlighted in blue/green may be used for searching. </li>
<li>They may be searched individually or in combination. To include data in a tab in a search: click on the tab to expand it and enter text to be searched. </li>
<li>When more than one field is used in a search, it is treated as an AND search. </li>
<li>Only the first gene or relationship entered for each category (MI, EC, DC) will be searched. </li>
<li>For Driver Components (DC), if you want to search for a non-mouse driver you need to select the organism before entering the gene.</li>
</ul>

<h4>Wildcards</h4>
<ul>
	<li>Wildcards (%) may be used for most fields. </li>
	<li>A wildcard may be used at the beginning, middle and/or end of your search text: e.g. %paired%, casein% or AK32%43%.</li>
</ul> 
 
<h4>Search by Dates</h4>
<ul>
	<li>Valid date formats for searching are:
		<ul>
			<li>mm/dd/yyyy or mm-dd-yyyy</li>
			<li>mm/dd/yy or mm-dd-yy</li>
			<li>yyyy/mm/dd or yyyy-mm-dd</li>
		</ul>
	</li>
	<li>Searching for a range of dates:
		<ul>
			<li>equal to: enter a single date</li>
			<li>greater than: &gt; mm/dd/yyyy</li>
			<li>greater than or equal to: &gt;= mm/dd/yyyy</li>
			<li>less than: &lt; mm/dd/yyyy</li>
			<li>less than or equal to: &lt;= mm/dd/yyyy</li>
			<li>between two dates (inclusive): mm/dd/yyyy..mm/dd/yyyy</li>
			</ul>
	</li>
</ul> 
<h4>Search Data</h4>
<p>When you click on the Search button:</p>
	<ul>
	<li>the results are displayed in a table with each row representing a single allele.</li>
	<li>Results are displayed by status (in progress, deleted, and reserved prior to approved or autoloaded) and in alpha-numeric order.</li>
	</ul>
<h3>Search by Allele</h3>
<p>
Entering an allele symbol alone or allele ID and tabbing out will automatically search for that allele. If you put a % after the allele symbol, you need to click search to start the search.
</p>
<h3>Marker Region Tool</h3>
<p>
You can not search for allele relationship annotations using the fields in the Marker Region Tool. This tool is only used to find markers to add to the Mutation Involves table.
</p>
<h3>Add New Mutation Involves Relationship</h3>
<ul>
	<li>Fields required to add a new relationship:
		<ul>
			<li>Must have an allele selected</li>
			<li>Relationship type</li>
			<li>J#</li>
			<li>Marker Acc ID</li>
			<li>Marker</li>
		</ul>
</ul>
<ul>
	<li>Autofill behavior:
		<ul>
			<li>Entering a Marker ID or Marker symbol will autofill the converse field</li>
			<li>Entering a J# autofills Citation</li>
			<li>If you want to change the marker once a marker has been entered you need to enter the MGI ID of the new marker. The ID field has precedence over the symbol field</li>
		</ul>
</ul>
<ul>
	<li>Duplicates:
		<ul>
			<li>The PWI will not enter duplicate &#40;identical marker and relationship type pair&#41;</li>
			<li>If you attempt to manually enter a duplicate and save the relationships the duplicate will not be entered and currently no warning is given</li>
			<li>The Marker Region tool does not include duplicates in the results set</li>
		</ul>
</ul>
<h3>Marker Region Tool</h3>
<p>
Allows for bulk addition of markers to the Mutation Involves table.
</p>
<ul>
	<li>Fields required:
		<ul>
			<li>Chr</li>
			<li>Start coordinate &#40;bp&#41;</li>
			<li>End coordinate &#40;bp&#41;</li>
			<li>Relationship type</li>
			<li>J#</li>
		</ul>
</ul>
<ul>
	<li>Markers included in the search results:
		<ul>
			<li>Descendants of marker types –</li>
			<ul>
				<li>protein coding gene</li>
				<li>non-coding RNA gene</li>
				<li>unclassified gene</li>
				<li>gene segment</li>
				<li>pseudogenic region</li>
			</ul>
			<li>Only markers NOT already associated to the allele with the specified relationship type are included in the count of results –</li>
			<ul>
				<li>Duplicate markers are skipped and not added</li>
			</ul>
		</ul>
</ul>
<ul>
	<li>To bulk add markers:
		<ul>
			<li>After filling in the required fields, clicking “Search Marker Count” finds the non-duplicate markers in the region</li>
			<li>Check that the count of markers matches what you expect &#40;use the Marker Query form to confirm&#41;</li>
			<li>Double check that you have an allele selected before proceeding</li>
			<li>Click “Add Markers to Mutation Involves”</li>
			<li>Markers are added to the Mutation involves table –</li>
			<ul>
				<li>Using the selected Relationship type and J#</li>
				<li>In chromosomal order</li>
			</ul>
			<li>You must still click “Save Relationships, Properties” to save the marker additions</li>
		</ul>
</ul>
<p>
Clearing the page or switching to a new allele does not clear the parameters in the Marker region tool. This allows you to add sets of markers to multiple alleles without re-entering all the fields. The tool has a separate clear button that only clears the fields in the tool.
</p>
<p>
There may be a delay in result return when adding large sets of markers &#40;1,000+&#41;. There may also be a delay when loading an allele with a large set of annotated markers.
</p>
<p>
If you want to check the list of markers before clicking the “Add Markers to Mutation Involves” table, you can click on the “Show JSON package/API domain + other data” button below the tables.
</p>
<h3>Add New Expresses Component Relationship</h3>
<ul>
	<li>Fields required to add a new relationship:</li>
		<ul>
		<li>Must have an allele selected</li>
		<li>Relationship type</li>
		<li>J#</li>
		<li>Marker</li>
                    <ul>
                    <li>The marker must be entered in MGI to be entered. If the Marker is not in MGI and is not from a species loaded in the non-mouse entrezgene load, you can add the marker and NCBI ID, if known, in the Marker PWI</li>
                    </ul>
		<li>IDs are not required but should be entered if you have one</li>
		</ul>
        <li>Autofill behavior:</li>
		<ul>
                <li>For non-mouse, you need to select the organism before entering the ID or symbol</li>
                <li>For Alliance species you can enter MOD or NCBI ID but the display will default to the NCBI ID</li>
                <li>For non Alliance species you can enter NCBI IDs</li>
                <li>Entering a J# autofills Citation</li>
                <li>If you want to change the marker once a marker has been entered you need to enter the ID of the new marker. The ID field has precedence over the symbol field</li>
		</ul>
        <li>Duplicates:</li>
		<ul>
                <li>The PWI will not enter duplicate (identical marker and relationship type pair)</li>
                <li>If you attempt to manually enter a duplicate and save the relationships the duplicate will not be entered and currently no warning is given</li>
		</ul>
        <li>Species in the non-mouse entrezgene load</li>
		<ul>
                <li>Cattle</li>
                <li>Chicken</li>
                <li>Chimpanzee</li>
                <li>Dog</li>
                <li>Human</li>
                <li>Macaque, rhesus</li>
                <li>Rat</li>
                <li>Western clawed (X. tropicalis)</li>
                <li>Zebrafish</li>
		</ul>
</ul>

<h3>Add New Driver Component Relationship</h3>
<p>
<ul>
<li>Fields required to add a new relationship:</li>
<ul>
<li>Must have an allele selected
<li>J#
<li>Marker
        <ul>
        <li>The marker must be entered in MGI to be entered. If the Marker is not in MGI and is not from a species loaded in the non-mouse entrezgene load, you can add the marker and NCBI ID, if known, in the Marker PWI</li>
        </ul>
<li>IDs are not required but should be entered if you have one
</ul>

<li>Autofill behavior:</li>
<ul>
<li>For mouse genes, entering a Marker ID or Marker symbol will autofill the converse field and autofill the organism</li>
<li>For non-mouse, you need to select the organism before entering the ID or symbol</li>
        <ul>
        <li>For Alliance species you can enter MOD or NCBI ID but the display will default to the NCBI ID</li>
        <li>For non Alliance species you can enter NCBI IDs</li>
        </ul>
<li>Entering a J# autofills Citation</li>
<li>If you want to change the marker once a marker has been entered you need to enter the ID of the new marker. The ID field has precedence over the symbol field</li>
</ul>

<li>Duplicates:</li>
<ul>
<li>The PWI will not enter duplicate (identical marker and relationship type pair)</li>
<li>If you attempt to manually enter a duplicate and save the relationships the duplicate will not be entered and currently no warning is given</li>
</ul>

<li>Species in the entrezgene load</li>
<ul>
<li>Cattle</li>
<li>Chicken</li>
<li>Chimpanzee</li>
<li>Dog</li>
<li>Human</li>
<li>Macaque, rhesus</li>
<li>Rat</li>
<li>Western clawed (X. tropicalis)</li>
<li>Zebrafish</li>
</ul>
</ul>

<h3>Deleting a relationship</h3>
<p>
Click on the ‘x’ next to a row to delete the selected relationship. Relationships need to have been saved for deletion to work.
</p>

<h3>Dialogue warning boxes &#40;pop ups&#41; behavior</h3>
<p>
Attempting to save a relationship without all of the required fields will trigger an error message listing the full set of required fields.
</p>
<p>
Attempting to save an “Expresses Components” relationship of type “expresses mouse gene” with any property fields filled in will trigger an error.
Attempting to save an “Expresses Components” relationship of type “expresses an orthologous gene” without required property fields filled in will trigger an error. 
</p>
<p>
Attempting to save relationships without an allele selected will trigger an error message “Cannot modify if a record is not selected”.
</p>
<p>
Entering an invalid NCBI gene ID, HGNC, RGD, or ZFIN ID will trigger an error message “Invalid Gene ID: #######”. There is no check for WB, FB, or SGD IDs as these are not loaded in MGI.
</p>
<p>
Entering an invalid MGI marker ID or symbol or J# will trigger an error message.
</p>
<p>
Entering a region in the Marker Region Tool that does not contain any non-duplicate valid markers or is an invalid region will trigger an error “No Markers Available for this Allele: region information entered”.
</p>
 

</div> <!-- helpSectionWrapper -->
