<div id="helpSectionWrapper">

<h2>P-WI Genotype Editor Interface functionality</h2>

<h4>General Description</h4>
<p>
The Genotype module is a search, display and edit page all in one. A user must be logged in to edit but can search without logging in.

</p><h4>Keyboard Short-cuts</h4>
There are reported issues using these short-cuts with Mac/Firefox. 
<ul>
	<li>Search = Ctrl-Alt-S</li>
	<li>Clear = Ctrl-Alt-C</li>
	<li>Modify = Ctrl-Alt-M</li>
	<li>Previous = Ctrl-Alt-P</li>
	<li>Next = Ctrl-Alt-N</li>
	
</ul>

<h4>Color-coded Labels</h4>
<ul>
        <li>Thistle = search only</li>
        <li>Green = search & edit</li>
        <li>Blue = read only</li>
        <li>Yellow = current row</li>
</ul>

<h4>Supported Browsers</h4>
<ul>
	<li>Chrome and FireFox are supported on both PC and Mac computers.</li>
	<li>Safari is NOT supported for the PWI. There are features that do not work.</li>
	<li>There are also reported issues using the keyboard short-cuts with Firefox for Mac.</li>
</ul>

<h3>Check boxes in dialogue warning boxes</h3>
<ul>
	<li>Certain actions in this module will generate popup, verifying that you want to do the action.</li>
	<li>Depending on how quickly you are working and your browser (Firefox in particular), subsequent popups might include a check box asking if you want to "Prevent this page from creating additional dialogs."</li>
	<li>Do <b>NOT</b> click the check box because clicking Yes in the popup is required to allow the action to proceed (and if the box is not shown, you cannot click Yes).</li>
	<li>The only way to undo clicking the check box is to quit and relaunch the browser.</li>
	<li>Therefore, it is best to IGNORE this check box.</li>
</ul>
<h3>Genotypes are used for</h3>
<ul>
	<li>Disease Ontology (DO) annotations</li>
	<li>GXD assays</li>
	<li>HT index samples</li>
	<li>Mammalian Phenotype (MP) annotations</li>
	<li>Strain associations</li>
	<li>(Phenotype) image associations</li>
</ul>
<h4>Determining where a genotype is used</h4>
Image associations, when present are displayed in the Image Pane Associations table.
All other data associations are displayed in the Citation/Strain/HT Sample table.
<ul>
	<li>When the genotype contains at least one allele pair, associations are displayed by default.</li>
	<li>For Genotypes that contain no allele pairs, "Get Data Sets for Genotype" button must be clicked to display the associations.</li>
</ul>
The data sets return is sorted first by J#, followed by the nulls (HT index samples and strains).

<h4>Rules governing creation/modification</h4>
A genotype cannot be deleted if it is being used for any of the above. The data must be moved before it can be deleted.
A genotype identical to one already in the system (aka duplicates) cannot be created, either by addition or modification.
The marker of the allele used and marker entered must match or the genotype cannot be created.<br>
The state of the allele pairs (eg homozygous, heterozygous) must match the entered alleles or the genotype cannot be created.

<h4>Strain</h4>
Text entered in this field will undergo validation.
If the text entered into the strain does not already exist in the database (as either a standard or non-standard strain), you will be given a dialogue box asking if you want to proceed.
If you forget to tab out, Not specified will be assigned by default.
You cannot create genotypes using "private" strains.
Phenotype curators must verify all newly-created standard strains, so as a courtesy to them, please enter any relevant information in the Private Curatorial Note field. (This is not needed for non-standard strains.)

<h4>Markers</h4>
When creating genotypes, the curator does not need to fill this field in; if they add an allele and tab out, the marker (and chromosome) field will automatically be filled in.
Generally, marker entry is case-insensitive; upon validation, the case will be corrected. <b>Exception:</b> in the cases where there are two markers that only differ in case (eg. T and t; Gs and gs) the system assumes you are entering the case of the marker you want to use.
In some cases when entering cytogenetic markers (eg. translocations, inversions), the system will require that you specify the chromosome of the marker you want to use.

<h4>Allele pair ordering</h4>
When a genotype is created, the rows are sorted alphabetically by marker.
A curator can then manually re-order the rows by changing the row number; use Modify to save changes.
<h4>Conditionally Targeted</h4>
Default = no.
<h4>Genotype Exists As</h4>
Default = mouse line.
<h4>Clipboard</h4>
Each curator has their own clipboard in the database. This allows the clipboard to be persistent, i.e. geneotypes remain in the clipboard between sessions until the curator clears it.
You must be logged in to add items to the clipboard.
The clipboard sorts each time an item is added to it; alphabetically by strain, then by allele pairs.
<h4>Image pane associations</h4>
When adding/modifying image associations, there must be at most 1 primary image.
Only full size images of type = Phenotype should be used for these associations.
<h3>Searching</h3>
The search return is sorted by strain and then by allele pair, with those genotypes that lack allele pairs sorting before those that do have them. 
All fields that have green or purple labels may be used for searching.
Fields may be searched individually or in combination. 
When more than one field is used in a search, it is treated as an AND search.
Allele searches will search both allele pair fields (ie a search using the Allele 1 field will also return matches found in the Allele 3 field).
Entering markers or alleles on >1 line in the genotype table will do an AND search for the markers/alleles on each line.
When searching by J#, be sure to tab out of the field.
The search wild card is %.
<ul>
	<li>It may be used for most fields.</li>
	<li>It may be used at the beginning, middle and/or end of your search text: %C57BL%, C57BL% or C57%6%.</li>
	<li>When doing wild card searching for markers or alleles, the wild card can only be used on row 1 of the table.</li>
</ul>
Please note that large searches take a long time to clear in browser world; it might be less frustrating to just reload the page. (remember: items on the clipboard will persist.)
 
<h4>Search by Dates</h4>
<ul>
	<li>Valid date formats for searching are:
		<ul>
			<li>mm/dd/yyyy or mm-dd-yyyy</li>
			<li>mm/dd/yy or mm-dd-yy</li>
			<li>yyyy/mm/dd or yyyy-mm-dd</li>
		</ul>
	</li>
	<li>Searching for a range of dates:
		<ul>
			<li>equal to: enter a single date</li>
			<li>greater than: &gt; mm/dd/yyyy</li>
			<li>greater than or equal to: &gt;= mm/dd/yyyy</li>
			<li>less than: &lt; mm/dd/yyyy</li>
			<li>less than or equal to: &lt;= mm/dd/yyyy</li>
			<li>between two dates (inclusive): mm/dd/yyyy..mm/dd/yyyy</li>
			</ul>
	</li>
</ul> 
<h4>Further information</h4>
See <a tabindex="-1" href="http://mgiwiki/mediawiki/index.php/sw:EI_Genotype">Here</a>
 
</div> <!-- helpSectionWrapper -->
