<div id="helpSectionWrapper">

<h2>P-WI Assay Editor Interface functionality</h2>

<h4>General Description</h4>
<p>
The Assay module is a search, display and edit page all in one. A user must be logged in to edit but can search without logging in.
</p>

<h4>Keyboard Short-cuts</h4>
There are reported issues using these short-cuts with Mac/Firefox. 
<ul>
        <li>Add = Ctrl-Alt-A</li>
        <li>Search = Ctrl-Alt-S</li>
        <li>Clear All = Ctrl-Alt-C</li>
        <li>Modify = Ctrl-Alt-M</li>
        <li>Delete = Ctrl-Alt-D</li>
        <li>First = Ctrl-Alt-F</li>
        <li>Previous = Ctrl-Alt-P</li>
        <li>Next = Ctrl-Alt-N</li>
        <li>Last = Ctrl-Alt-L</li>
        <li>Move between Specimen and Results tables = Ctrl-Alt-R</li>
</ul>

<h4>Color-coded Labels</h4>
<ul>
        <li>Thistle = search only</li>
        <li>Green = search &amp; edit</li>
        <li>Blue = read only</li>
        <li>Yellow = current row</li>
</ul>

<h4>Supported Browsers</h4>
<ul>
        <li>Chrome and FireFox are supported on both PC and Mac computers.</li>
        <li>Safari is NOT supported for the PWI. There are features that do not work.</li>
</ul>

<h3>Searching</h3>
<p>
All fields that have labels highlighted in thistle or green may be used for searching. 
They may be searched individually or in combination. To include data in the assay note in a search: click on the button to expand it and enter text to be searched. When more than one field is used in a search, it is treated as an AND search. If you search by a probe ID without adding the marker, you can't tab through it. Add ID then click search. If you tab through, it will give you an error (because there is no gene to be in agreement) and the field will blank.
</p>

<h4>Search by Dates</h4>
<ul>
        <li>Valid date formats for searching are:
                <ul>
                        <li>mm/dd/yyyy or mm-dd-yyyy</li>
                        <li>mm/dd/yy or mm-dd-yy</li>
                        <li>yyyy/mm/dd or yyyy-mm-dd</li>
                </ul>
        </li>
        <li>Searching for a range of dates:</li>
                <ul>
                        <li>equal to: enter a single date</li>
                        <li>greater than: &gt; mm/dd/yyyy</li>
                        <li>greater than or equal to: &gt;= mm/dd/yyyy</li>
                        <li>less than: &lt; mm/dd/yyyy</li>
                        <li>less than or equal to: &lt;= mm/dd/yyyy</li>
                        <li>between two dates (inclusive): mm/dd/yyyy..mm/dd/yyyy</li>
                </ul>
        
</ul> 

<h3>Page organization</h3>
The page is divided into
<ul>
        <li>Marker/J# section – citation, gene, assay type, assay note</li>
        <li>Probe/Antibody section – accession ID, details of prep, label, visualization method</li>
                <ul>
                        <li>Section displayed depends on assay type or detection method</li>
                        <li>Reporter assays include Reporter gene and Detection method menus.</li>
                </ul>   
        <li>Specimen Table/Gel Lanes Table – specimen/lane label, genotype, age, sex, lane or specimen note</li>
                <ul>
                        <li>in situ-type assays have fix, embedding, hybridization, and the DL checkbox </li>
                        <li>blot assays include structure, sample amount, RNA type</li>
                </ul> 
		<li>Action button row - see below for descriptions</li>
				<ul>
                        <li>Double Label - only in situ-type assays</li>
                        <li>Add Assay Structures to EMAPA Clipboard</li>
						<li>Add Assay Cells to Cell Type Clipboard - only in situ-type assays</li>
						<li>Add Assay Genotype to Genotype Clipboard</li>
						<li>Replace Genotype</li>
                </ul> 
        <li>Results Table or Gel Rows &amp; Bands Table</li>
                <ul>
                        <li>Structure(s), cell type(s), strength, pattern, image panes, results note for in situ assays</li>
                        <li>Size, units, Row note, Band strength, Band notes for blot assays</li>
                </ul>   
</ul>

<h3>Assay Types</h3>
<h4>In situ types</h4>
<ul>
        <li>Immunohistochemistry</li>
        <li>Reporter</li>
                <ul>
                        <li>Knock in (GXD &amp; Cre)</li>
                        <li>Transgenic (Cre)</li>
                        <li>Recombinase (Cre)</li>
                </ul>   
        <li>RNA in situ</li>
</ul>

<h4>Blot types</h4>
<ul>
        <li>Northern (probe)</li>
        <li>RT-PCR (primer)</li>
        <li>Western (antibody)</li>
        <li>S1 nuclease (legacy data)</li>
        <li>RNase protection (legacy data)</li>
</ul>
<p>
Curators can change an in situ type assay to another in situ type assay. Same for blots. However, you can’t make in situ type a blot type or vice versa.
</p>

<h3>All assays</h3>
<h4>Assay note</h4>
<ul>
        <li>default for note is closed.</li>
        <li>Click to open, focus goes into the field.</li>
        <li>Button is Green when has entry.</li>
        <li>After entry, tab takes you to assay type.</li>
</ul>

<h4>Search return</h4>
<ul>
        <li>Show ~10 rows before scroll</li>
        <li>Display is J#;assay type;gene</li>
        <li>sort order: </li>
                <ul>
                        <li>J#</li>
                        <li>Assay type</li>
                        <li>Gene</li>
                </ul>
        <li>Slim Vocab for assays in the search return</li>
                <table>
                        <tbody><tr>
                                <th>Abbreviation</th>
                                <th>Assay type
                                </th><th>Number
                        </th></tr><tr>
                                <td>Immuno</td>
                                <td>Immunohistochemistry</td>
                                <td>6</td>
                        </tr>
                        <tr>
                                <td>InSitu</td>
                                <td>RNA in situ</td>
                                <td>1</td>
                        </tr>
                        <tr>
                                <td>Knock</td>
                                <td>In situ reporter (knock in)</td>
                                <td>9</td>
                        </tr>
                        <tr>
                                <td>North</td>
                                <td>Northern blot</td>
                                <td>2</td>
                        </tr>
                        <tr>
                                <td>Recomb</td>
                                <td>Recombinase reporter</td>
                                <td>11</td>
                        </tr>
                        <tr>
                                <td>RNAse</td>
                                <td>RNase protection</td>
                                <td>4</td>
                        </tr>
                        <tr>
                                <td>RTPCR</td>
                                <td>RT-PCR</td>
                                <td>5</td>
                        </tr>
                        <tr>
                                <td>S1Nuc</td>
                                <td>Nuclease S1</td>
                                <td>3</td>
                        </tr>   
                        <tr>
                                <td>Transg</td>
                                <td>In situ reporter (transgenic)</td>
                                <td>10</td>
                        </tr>
                        <tr>
                                <td>West</td>
                                <td>Western blot</td>
                                <td>8</td>
                        </tr>
                </tbody></table>
</ul>

<h4>Duplicate Prep</h4>
<ul>
        <li>Placement right justified in J#/citation row</li>
        <li>New assay/new MGI:ID</li>
        <li>created by user/date</li>
        <li>The following data from the previous assay</li>
                <ul>
                        <li>J# &amp; citation</li>
                        <li>Assay notes</li>
                        <li>Assay type</li>
                        <li>Marker</li>
                        <li>Probe prep of the previous assay (e.g. probe/antibody ID, name, labeled with etc.)</li>
                        <li>Specimen/Results or Gel Lanes/Rows &amp; Bands tables are blank</li>
                </ul>
</ul>

<h4>Duplicate Partial</h4>
<ul>
        <li>Placement right justified in Assay Type/Marker row next to the Duplicate 'All' button.</li>
        <li>Same as duplicate prep plus...</li>
        <li>Entire Specimen or Gel Lanes table (all rows and notes)</li>
        <li>Results or Gel Rows &amp; Bands table is blank</li>
</ul>

<h4>Duplicate All</h4>
<ul>
        <li>Placement right justified in Assay Type/Marker row next to the Duplicate 'Partial' button.</li>
        <li>Same as duplicate partial plus...</li>
        <li>Entire Results or Gel Rows &amp; Bands table including all structures, images, and notes in all rows (if present; not required for duplication).</li>
</ul>

<h4>Minimum width fields</h4>
When user adjusts the width of browser window, a minimum width of this field/column prevents width from shrinking below a set point. (not the same as column resizing - see section below)
<ul>
        <li>Genotype field - shows full id</li>
        <li>MGI ID field - shows full id</li>
</ul>

<h4>Pulldown Menus</h4>
<ul>
                <li>Pulldown preference - menus that list both Not Specified and Not Applicable will preferentially enter Not Specified when you start typing 'N'.</li>
                <li>You can use the arrow keys to move through the entries of the menu (left, right, up, down). Some keyboards will also open the menu using up, down arrows.</li>
</ul>

<h4>Standardized Age Notes</h4>
<ul>
<li>There are three buttons that can be used to add standardized notes to the Age Notes field. These buttons add the notes to all specimens/gel lanes in the assay.</li>
</ul>

<h4>Tab-through copying from the line above</h4>
<p>Once a line has been filled in, the fields in all subsequent rows can be copied by tabbing-through in the table.</p>
<ul>Exceptions
        <li>Age Notes</li>
        <li>Specimen Notes (in situs)</li>
        <li>R (in situs)</li>
        <li>Lane Notes (blots)</li>
        <li>Band Strength (blots)</li>
        <li>Band Notes (blots)</li>
        <li>EMAPA (in situs)</li>
                <li>CL (in situs)</li>
        <li>Results Notes (in situs)</li>
        <li>Control Lane Label (blots)</li>
</ul>

<h4>Re-ordering Rows</h4>
<p>The # column lets the curator set the row # in case they need to be re-ordered. Often due to clustering the initial entry to use tab-through copy for genotypes or ages, etc.</p>

<h4>Error messages</h4>
<ul>
        <li>Delete: You will be given a popup verifying that you want to do this.</li>
        <li>Required Field : Assay Type</li>
        <li>Required Field : Probe Prep (where applicable)</li>
        <li>Required Field : Antibody Prep (where applicable)</li>
        <li>Delete table rows that haven't been saved: Postgres errors.</li>
        <li>Postgres error for Gel Row size with no unit</li>
        <li>Message: There is unsaved data in the current Assay</li>
        <li>Postgres error for Gel Lane with no Control value.</li>
        <li>Incomplete (red inline text in blots to indicate unsaved gel bands)</li>
</ul>

<h3>Column Controls</h3>
<p>Column settings will be maintained with 'Clear' and when switching to other assays in the search return. To fully reset, you reload the page. This is preferred for curation/QC.</p>

<h4>Column Resizing</h4>
<ul>
                <li>Specimen table: Specimen Label/Genotype/Age Range : horizontal only</li>
                <li>Specimen table: Age Note/Specimen Note: horizontal &amp; vertical</li>
                <li>Results table: Image Pane : horizontal only</li>
                <li>Results table: Note : horizontal &amp; vertical</li>
                <li>Lane table: Lane label/Genotype/Sample Amt/Age Range : horizontal only</li>
                <li>Lane table: Age Note/Lane Note: horizontal &amp; vertical</li>
                <li>Gel Rows &amp; Bands table: Row Note/Band Note: horizontal &amp; vertical</li>
                <li>Use text corner drag to set to column width desired.</li>
                <li>Resizing does not apply to menu/pulldown columns.</li>
                <li>Resizing works on any row in the table.</li>
</ul>

<h4>Copy column</h4>
<p>An icon in header (square with a C) supports the ability to copy the value in a cell in the specimen, results, or gel lanes table to all other cells in that column (only rows with a label are saved). A very useful aspect of this feature is using it from a blank field (including note fields) to clear an entire column. Copy column functionality has been removed from the specimen and gel lane label columns.</p>

<h4>Hidden Columns</h4>
<ul>An icon in the header (circle with an H) supports hide/show the column
                <li>When the column is hidden, it will not completely disappear. The first 2 letters of the header and the icon will be visible.</li>
                <li>Defaults will be entered &amp; age notes buttons work on the hidden columns.</li>
                <li>Hovering over the header shows a popup of what column it is.</li>
                <li>If you have columns hidden, yellow button above the table will 'Show All Columns'</li>
                <li>Pulldown columns will be hideable, but not resizeable.</li>
</ul>

<h4>Frozen Columns</h4>
<ul>When there is a horizontal scroll bar, these columns remain visible at all times:
                <li>Specimen (in situ type assays Specimen Table)</li>
                <li>Lane (blot type assays Gel Lane Table)</li>
                <li>Sidebar listing the lane labels (blot type assays Gel Rows and Bands table)</li>
</ul>

<h4>Sort Column</h4>
<p>An icon in the header (up-down arrowheads) supports the ability to sort the column entries alphanumerically. The sort change is not saved.</p>

<h3>Clipboards</h3>
<h4>Genotype Clipboard</h4>
<ul>
        <li>Placement - On right side below search return list.</li>
        <li>update button in the header for changes we make in the genotype module</li>
        <li>Scroll bar for longer lists – depends on the genotypes/strains chosen, but ~8 rows</li>
        <li>Clipboard size is not adjustable.</li>
        <li>Validation works with MGI:#### and ##### typed into genotype field of specimen table.</li>
        <li>The "genotype display" is a combination of:</li>
                <ul>
                        <li>Strain,</li>
                        <li>Allele1,Allele2</li>
                </ul>   
        <li>yellow highlight of the genotype clipboard entry for the current specimen row</li>
        <li>Clipboard row should be visible - scrollIntoView</li>
        <li>Selecting a clipboard entry inserts that ID into the genotype field of the selected specimen row. Either adds to a blank or replaces an entry that already exists.</li>
        <li>Deselecting a clipboard entry does nothing.</li>
        <li>Sort in genotype clipboard:</li>
                <ul>
                        <li>First sort is by Used/Unused in the current assay.</li>
                                <ul>
                                        <li>Used = [MGI:###]</li>
                                        <li>Unused = *[MGI:###]</li>
                                </ul>   
                        <li>Within those groups sort is by MGI ID of genotype.</li>
                        <li>If a new genotype module - genotype is added, and you refresh the assay module - genotype clipboard, the new will appear in the correct unused set of genotypes by ID #.</li>
                </ul>
        <li>Can only add one genotype to a specimen</li>
</ul>

<h4>Image Clipboard</h4>
<ul>
        <li>Placement - On left side of Results table.</li>
        <li>Show 6 rows before scroll (in situ-type assays)</li>
		<li>Show 3 rows before scroll (blot-type assays)</li>
        <li>Size of the clipboard can be made shorter or longer, thinner & wider</li>
        <li>Select/Add to Results</li>
        <li>De-Select removes the image pane</li>
        <li>Blot type assays - only 1 image pane allowed; In situ type assays - multiple image panes allowed on a single result row.</li>
        <li>Image Panes : xxx result(s) : click will refresh/resort clipboard from the image module</li>
                <ul>
                        <li>sort order is alpanumeric.</li>
                        <li>if Image Pane(s) is being used, then visible – scrollIntoView.</li>
                        <li>Entries hightlighted yellow.</li>
                        <li>the used image panes stay in place on the list - don't float to the top of the list.</li>
                        <li>lists all image panes of the J#</li>
                </ul>
</ul>

<h4>EMAPA Clipboard</h4>
<ul>
        <li>Placement - on left side of Results table below the Image panes clipboard.</li>
        <li>Scroll bar appears for long lists.</li>
        <li>Clipboard header: EMAPA stage;structure: # result(s). Click header to refresh both the EMAPA and CellType clipboards with changes made in their respective modules.</li>
        <li>Clipboard entry format is TS#;structure</li>
        <li>Select/Add structure to Results</li>
        <li>De-Select removes the structure from the results</li>
        <li>Can add multiple structures to a single result.</li>
        <li>First Sort is:</li>
                <ul>
                        <li>Used = TS#;structure name</li>
                        <li>Unused = *TS#;structure name</li>
                </ul>
        <li>Sort within Used Structures: </li>
                <ul>
                        <li>All structures used in all the occupied rows of the results table are considered Used. See MGI:5421193</li>
                        <li>The structure associated with the selected row of the results table is highlighted yellow within the list of Used Structures in the clipboard.</li>
                        <li>The order of the Used structures is by Results table row # (not alphanumeric). See MGI:5421193</li>
                        <li>If multiple structures are used on a single row those structures are grouped in the clipboard and are in alphanumeric order.see MGI:6356718 Specimen 1A</li>
                </ul>
        <li>Sort within Unused Structures is determined by their order in the EMAPA module-clipboard.</li>
</ul>

<h4>Cell Ontology (Cell Types) Clipboard</h4>
<ul>
        <li>Placement - below Results table.</li>
        <li>Scroll bar appears for long structure lists.</li>
        <li>Clipboard header: Cell Ontology: # result(s) Click header to refresh both the EMAPA and CellType clipboards with c    hanges made in their respective modules.</li>
        <li>Select/Add cell type to CL field in Results table (in situ-type assays only)</li>
        <li>De-Select removes the cell type from the results table</li>
        <li>Can add multiple cell types to a single result row.</li>
        <li>First Sort is:</li>
                <ul>
                        <li>Used = cell type</li>
                        <li>Unused = *cell type</li>
                </ul>
        <li>Sort within Used Structures: </li>
                <ul>
                        <li>All cell types used in all the occupied rows of the results table are considered Used.</li>
                        <li>The cell type associated with the selected row of the results table is highlighted yellow within the list of Used cell types in the clipboard.</li>
                        <li>The order of the Used cell types is by Results table row # (not alphanumeric).</li>
                        <li>If multiple cell types are used on a single row they are grouped in the clipboard and are in alphanumeric order.</li>
                </ul>
        <li>Sort within Unused Structures is determined by their order in the Cell Type Browser-clipboard.</li>
</ul>

<p>
The 'Add assay struct to EMAPA clipboard' button under the Specimen/Gel Lanes Table populates the EMAPA clipboard in the EMAPA module with all structures associated with all specimens or blot lanes in the current assay.
</p>
<p>
The 'Add Assay Cells to Cell Type Clipboard’ button under the Specimen Table populates the CL clipboard in the CL module with all cell types associated with all specimens in the current assay.  Note: the CL module clipboard does not have a refresh button, so use Sort to see the newly added cell types.
</p>
<p>
The 'Add Assay Genotype to Genotype Clipboard' button under the Specimen/Gel Lanes Table populates the Genotype clipboard in the Genotype module with all genotypes associated with all specimens or blot lanes in the current assay.
</p>

<h3>In situ assay-specific details</h3>
<h4>Defaults</h4>
If adding new Specimen row and Specimen Label is not blank, then if left blank these defaults are set:
<ul>
        <li>Genotype = MGI:2166310</li>
        <li>Age = embryonic day</li>
        <li>Sex = not specified</li>
        <li>fixation = not specified</li>
        <li>embedding = not specified</li>
        <li>hybridization = not specified</li>
</ul>

Although the display of some vocabulary terms can be modified/abbreviations changed, DO NOT modify the terms in the GXD Hybridization vocabulary; BAD THINGS WILL HAPPEN.

<h4>Specimen table</h4>
<p>
Default number of rows in the specimen table overall should be 24. (before you have to click to add more rows) Number to show before a scroll bar = 7. After ADD or Modify, the table should contain 10 additional blank rows.
</p>

<h4>Double Label Tool</h4>
<p>When a specimen has been double labeled, we add a note that indicates the colors of each gene or other element with links to the other assay for the user. The Double Label Tool is a one-way editing tool to allow curators to build complex notes with gene symbols, ID's and label colors.</p>
<ul><b>Note templates in the DL tool</b>
		
        <li>Double labeled: color1 - gene1; color2 - gene2 (assay \Acc(*||)).</li>
		<li>Double labeled: color1 - gene1 extraword; color2 - gene2 (assay \Acc(*||))</li>
        <li>Double labeled: color1 - geneX extraword; color2 - geneX extraword (assay \Acc(*||))</li>
		<li>Triple label all coding sets. Triple labeled: color1 - gene1; color2 - gene2 (assay \Acc(*||)); color3 - gene3 (assay \Acc(*||)).</li>
        <li>Multi-labeled. Multi-labeled: color1 - gene1; color2 - gene2 (assay \Acc(*||)); color3 - gene3 (assay \Acc(*||)); color4 - gene4 (assay \Acc(*||)).</li>
		<li>Double labeled: color1 - gene1 and gene2; color3 - gene3 (assay \Acc(*||)).</li>
        <li>Double labeled: color1 - gene1; color2 - gene2 (assay \Acc(*||)) and gene3 (assay \Acc(*||)).</li>
		<li>Multi-labeled with Coding and text+color sets. Multi-labeled: color1 - gene1; color2 - gene2 (assay \Acc(*||)); color3 - gene3 (assay \Acc(*||)); color4 - text4.</li>
		<li>Double labeled: color1 - gene1; color2 - text2</li>
</ul>
		
<ul><b>Basic Directions</b>
        <li>Use the DL checkbox in the specimen table to select rows (required)</li>
        <li>Open the Double Label tool using the button</li>
        <li>Add the colors for the specimen (first color column)</li>
        <li>Add the colors for the other genes</li>
		<li>Click the Text+color to add those 2 columns for any label or stain that doesn't have an assay ID</li>
        <li>Click preview button at the top of the table to see it in the preview text field.</li>
        <li>Click Process to add the note to the specimen note, and it will also modify the assay.</li>
</ul>		
<p>The tool knows what genes to display (and their assay ID's) based on the image panes attached to the specimens. If the gene displayed is incorrect, check the image pane summary.</p>

<ul>Other functions in the DL tool       
		<li>Clear - sets table to blank; refreshes the table based on changes to checkboxes</li>
        <li>Copy column - in color picklist and text headers  </li>
		<li>Preview - click once to show the note prior to processing. Press again to 'refresh' the note based on rules/templates. Will overwrite any changes made in 'preview'.</li>
        <li>double label note field (aka 'preview note') is editable; changes will be saved to specimen note upon processing.</li>
		<li>Process adds note to specimen field, modifies assay, and closes the table.</li>
		<li>Entries in the DL tool are not objects in the database - reopening the tool will not re-open the existing specimen note.
</ul>

<p>There are two buttons (Add Double Label; Add Acc/MGI) that can also be used as blanks in the specimen table directly; replace the * in the note with the MGI id of the assay you want to create the link to the other assay.</p>

<h4>Results table</h4>
<ul>
		<li>Default number of rows in the results table overall should be 4. (before you have to click to add more rows) Number to show before a scroll bar = 14. After ADD or Modify, the table should contain 4 additional blank rows.</li>
		<li># column - curator can change the order of the rows in the results table. Field is not part of tabbing. Column is hidden by default - open column using circled H in header.</li>
		<li>CL column is not part of tabbing</li>
		<li>Next, Current, Prev buttons</li>
			<ul>
                <li>When adding results, use the 'Next' button above the table to advance to the next Specimen without having to click into the Specimen table.</li>
                <li>Use 'Prev' to go back one specimen. </li>
                <li>Use 'Current' to move cursor focus to the current Specimen in the Specimen table.</li>
			</ul>	
</ul>

<h3>Blot assays - specific details</h3>
<h4>Defaults</h4>
A 'used/new' Gel Lane is one that has a Lane label. If only Lane filled in on Add or Modify, then set these defaults:
<ul>
                <li>Control = No</li>
                                <li>Genotype = MGI:2166310</li>
                <li>Westerns only set RNA = Not Applicable; All other blots set RNA = Not Specified</li>
                <li>Age = embryonic day</li>
                <li>Sex = Not Specified</li>
</ul>

If new Gel Lane row and Control = anything other than ‘No’, then set these defaults:
<ul>
                <li>Genotype = MGI:2166309</li>
                <li>RNA=Not Applicable</li>
                <li>Age=Not Applicable</li>
                <li>Sex=Not Applicable</li>
</ul>

For all Used Rows of the Gel Lanes Table update the Gel Rows and Bands Table:
<ul>
                <li>Each Lane should have a Lane #;Label and Band Note in the Gel Rows and Bands table</li>
                <li>If Control != ‘No’, then set Gel Band Strength = Not Applicable.</li>
                <li>Default Unit = Not Specified</li>
</ul>

<h4>Lanes table</h4>
<p>
Default number of rows in the lanes table should be 20. (before you have to click to add more rows) Number to show before a scroll bar = 10. After ADD or Modify, the table should contain 1 additional blank row.
</p>

<h4>Gel Rows and Bands table</h4>
<ul>
                <li>Gel Rows &amp; Bands shows all Bands for the Lanes in the assay (no scroll).</li>
                <li>Multiple Gel Rows in the Gel Rows &amp; Bands Table automatically sort descending by size on Modify.</li>
                                <li>When adding additional Gel Rows:</li>
                                        <ul>
                                                <li>When you add a Gel Row 2 or more, the size, unit, and all strengths fields are blank (even for Control !=No lanes).</li>
                                                <li>Curator must add all Control=No (data) lane strengths prior to modify.</li>
                                                <li>After Modify, Control != No lanes will autofill the Not Applicable.</li>
                                                <li>If strength is not added before modify, the strength menu will disappear for all data lanes, and the row will have to be deleted and re-added.</li>
                                        </ul>
</ul>

<h3>Replace genotype</h3>
In all assay modules. This finds all specimens/gel rows using a particular genotype for a J#, then replaces it for all assays (not just the current selected one). Yellow button below the Results Table.
<ul>
        <li>Type J# tab to Current genotype acc ID field</li>
        <li>type current genotype acc ID, tab</li>
        <li>type new genotype acc ID, tab</li>
        <li>Process. Updates assay record and expression cache.</li>
</ul>

<h4>JSON</h4>
<ul>Process status
                <li>C - create</li>
                <li>d - delete</li>
                <li>u - update</li>
                <li>x - exists, no changes</li>
</ul>
<h3>Further Information</h3>
<ul>
<li><a href="http://mgiprodwiki.jax.org/mediawiki/index.php/sw:EI_Assay_Module">sw:EI_Assay_Module</a></li>
<li><a href="http://mgiprodwiki.jax.org/mediawiki/index.php/GXD#tab=Documentation">GXD documentation</a></li>
</ul>

</div> <!-- helpSectionWrapper -->
