<div id="helpSectionWrapper">

<h2>P-WI Genotype Editor Interface functionality</h2>

<h4>General Description</h4>
<p>
The Genotype module is a search, display and edit page all in one. A user must be logged in to edit but can search without logging in.

</p><h4>Keyboard Short-cuts</h4>
Shortcuts are used for creating, updating, deleting and searching genotypes. 
<ul>
	<li>Search = Ctrl-Alt-S</li>
	<li>Clear = Ctrl-Alt-C</li>
	<li>Modify = Ctrl-Alt-M</li>
	<li>Add = Ctrl-Alt-A</li>
	<li>Delete = Ctrl-Alt-D
	<li>Next = Ctrl-Alt-N</li>
	<li>Previous = Ctrl-Alt-P</li>
	<li>Last = Ctrl-Alt-L</li>	
	<li>Genotype Clipboard Add = Ctrl-Alt-G</li>
</ul>
There are no keyboard shortcuts for the genotypes clipboard.

<h4>Color coding for Field Labels</h4>
<ul>
        <li>Thistle = search only</li>
        <li>Green = search & edit</li>
        <li>Blue = Read-only fields</li>
        <li>Yellow = current row</li>
</ul>

<h4>Supported Browsers</h4>
<ul>
	<li>Chrome and FireFox are supported on both PC and Mac computers.</li>
	<li>Safari is NOT supported for the PWI. There are features that do not work.</li>
	<li>There are also reported issues using the keyboard short-cuts with Firefox for Mac.</li>
</ul>

<h4>Check boxes in dialogue warning boxes</h4>
<ul>
	<li>Certain actions in this module will generate popup, verifying that you want to do the action.</li>
	<li>Depending on how quickly you are working and your browser (Firefox in particular), subsequent popups might include a check box asking if you want to "Prevent this page from creating additional dialogs."</li>
	<li>Do <b>NOT</b> click the check box because clicking Yes in the popup is required to allow the action to proceed (and if the box is not shown, you cannot click Yes).</li>
	<li>The only way to undo clicking the check box is to quit and relaunch the browser.</li>
	<li>Therefore, it is best to IGNORE this check box.</li>
</ul>
<h4>Genotypes are used for</h4>
<ul>
	<li>Disease Ontology (DO) annotations</li>
	<li>GXD assays</li>
	<li>HT index samples</li>
	<li>Mammalian Phenotype (MP) annotations</li>
	<li>Strain associations</li>
	<li>(Phenotype) image associations</li>
</ul>
<h4>Determining where a genotype is used</h4>
<ul>
	<li>Image associations, when present are displayed in the Image Pane Associations table.</li>
	<li>All other data associations are displayed in the Citation/Strain/HT Sample table.
		<ul>
			<li>When the genotype contains at least one allele pair, associations are displayed by default.</li>
			<li>For Genotypes that contain no allele pairs, "Get Data Sets for Genotype" button must be clicked to display the associations.</li>
			<li>The data sets return is sorted first by J#, followed by the nulls (HT index samples and strains).</li>
		</ul>
	</li>
</ul>
<h4>Rules governing creation/modification/deletion</h4>
<ul>
	<li>A genotype cannot be deleted if it is being used for any of the above (aka referential integrity). The data must be moved before it can be deleted.</li>
	<li>A genotype identical to one already in the system (aka duplicates) cannot be created, either by addition or modification.</li>
	<li>The marker of the allele used and marker entered must match or the genotype cannot be created.</li>
	<li>The state of the allele pairs (eg homozygous, heterozygous) must match the entered alleles or the genotype cannot be created.</li>
</ul>

<h4>Creation via Deselection</h4>
<p>This module has been enabled to allow creation of genotypes using as a starting point one that has been "deselected" on the search return. This feature is useful when creating related genotypes, such as het/homs on the same strain background. To do this:
<ul>
	<li>Have the genotype you want to use as your starting point displayed.</li>
	<li>Click on it in the search return, so it's no longer yellow; its MGI ID will be blanked out and the notes fields will clear.</li>
	<li>Modify values to what you want; do NOT add any notes yet.</li>
	<li>Hit Add to create genotype.</li>
	<li>Now you can add general and private curatorial notes, if needed.</li>
</ul>
<p>NOTE: A special case exists for entering repulsion compound heterozygotes.<br>
To enter repulsion compound heteroheterozygous genotypes (eg. Dock7&lt;+&gt; Lepr&lt;db&gt;/Dock7&lt;m&gt; Lepr&lt;+&gt;), enter each allele separately in the Allele 1 field, leave Allele 2 blank and use the Hemizygous Insertion state. Use the Compound attribute to determine the placement of the allele. </p>
<p>In the above example:
<ol>
	<li>Line #1 would have Dock7&lt;+&gt; as Allele 1, Hemizygous insertion for State, and Top for Compound.</li>
	<li>Line #2 would have Lepr&lt;db&gt; as Allele 1, Hemizygous insertion for State, and Top for Compound.</li>
	<li>Line #3 would have Dock7&lt;m&gt; as Allele 1, Hemizygous insertion for State, and Bottom for Compound.</li>
	<li>Line #4 would have Lepr&lt;+&gt; as Allele 1, Hemizygous insertion for State, and Bottom for Compound.</li>
</ol>
<h4>Genetic Background/Strain</h4>
<ul>
	<li>Text entered in this field will undergo validation.</li>
	<li>If the text entered into the genetic background/strain field does not already exist in the database (as either a standard or non-standard strain), you will be given a dialogue box asking if you want to proceed.</li>
	<li>If you forget to tab out of this field, Not specified will be assigned by default.</li>
	<li>You cannot create genotypes using "private" strains.</li>
	<li>Phenotype curators must verify all newly-created standard strains, so as a courtesy to them, please enter any relevant information in the Private Curatorial Note field. (This is not needed for non-standard strains.)</li>
</ul> 
 
<h4>Markers</h4>
<ul>
	<li>When creating genotypes, the curator does not need to fill this field in; if they add an allele and tab out, the marker (and chromosome) field will automatically be filled in.</li>
	<li>Generally, marker entry is case-insensitive; upon validation, the case will be corrected. <b>Exception:</b> in the cases where there are two markers that only differ in case (eg. T and t; Gs and gs) the system assumes you are entering the case of the marker you want to use.</li>
	<li>In some cases when entering cytogenetic markers (eg. translocations, inversions), the system will require that you specify the chromosome of the marker you want to use.</li>
</ul>

<h4>Allele pair ordering</h4>
When a genotype is created, the rows are sorted alphabetically by marker.
A curator can then manually re-order the rows by changing the row number; use Modify to save changes.
<h4>Conditionally Targeted</h4>
Default = no.
<h4>Genotype Exists As</h4>
Default = mouse line.

<h4>Clipboard</h4>
<ul>
<li>Each curator has their own clipboard in the database. This allows the clipboard to be persistent, i.e. geneotypes remain in the clipboard between sessions until the curator clears it.</li>
<li>You must be logged in to add items to the clipboard.</li>
<li>The clipboard sorts each time an item is added to it; alphabetically by strain, then by allele pairs.</li>
</ul>

<h4>Genotype Summary for Genotype Module Search Results Functionality:</h4>
<ul>
<li>User enters parameters, then clicks Search Summary button. The Search summary button should not be used after clicking Search (use one or the other).</li>
<li>All genotypes listed on the search results box are included in the summary.</li>
<li>Top of the summary lists the MGI ID's for all records in the search results.</li>
<li>Sort is by genetic background, then by allele detail display. Due to different ways the API compiles the data for the search results versus the summary, entries with multiple alleles may not match the sort of the results.</li>
<li>Yes’es in the Annotations columns are based on any annotations using the genotype, not solely those derived from the reference.</li>
<li>The summary logic does not use the expression cache. This means genotypes appear on the summary once they are annotated to specimens/gel lanes; it is not necessary that the results are annotated.</li>
<li>The URL supports /pwi/edit/genotypesummary/?accid=MGI:#######,MGI:######,etc.</li>
</ul>   

<h4>Genotype Summary for Clipboard Functionality:</h4>
<ul>
<li>All genotypes listed on the user's clipboard are included.</li>
<li>Sorted as on the clipboard.</li>
<li>Yes’es in the Annotations columns are based on any annotations using the genotype, not solely those derived from the reference.</li>
<li>The summary logic does not use the expression cache. This means genotypes appear on the summary once they are annotated to specimens/gel lanes; it is not necessary that the results are annotated.</li>
<li>The URL supports /pwi/edit/genotypesummary/?user_id=username</li>
</ul>   

<h4>Pulldown Menus</h4>
Genotype: State, Compound<br>
MP Annot: Qualifier, Evidence, Sex<br>
DO Annot: Qualifier Evidence<br>
GO Annot: Qualifier Evidence<br>
<p>You can always use the pulldown to select the item you want to use for annotation. </p>
<p>Fields that have pulldowns you can tab into --State and Compound-- have dropdown focus:
<ul>
	<li>Start typing; the field will fill in an item that matches what you have typed.</li>
	<li>If you keep typing, it will change the fill to match what is typed.</li>
	<li>Alternatively, you can use the space bar to show all options; use the arrow keys to get to the desired item and tab out to select.</li>
</ul>
	
<h4>Image pane associations</h4>
When adding/modifying image associations, there must be at most 1 primary image.
Only full size images of type = Phenotype should be used for these associations.
<h4>Searching</h4>
<ul>
	<li>The search return is sorted by strain and then by allele pair, with those genotypes that lack allele pairs sorting before those that do have them.</li> 
	<li>All fields that have green or purple labels may be used for searching.</li>
	<li>Fields may be searched individually or in combination.</li> 
	<li>When more than one field is used in a search, it is generally treated as an AND search. Exception: when allele searches use both allele fields on the same table row, it is treated as an OR search.</li>
	<li>Allele searches will search both allele pair fields (ie a search using the Allele 1 field will also return matches found in the Allele 2 field and vice versa).</li>
	<li>Entering markers or alleles on >1 line in the genotype table will do an AND search for the markers/alleles on each line.</li>
	<li>When searching by J#, be sure to tab out of the field.</li>
	<li>The search wild card is %.
		<ul>
			<li>It may be used for most fields.</li>
			<li>It may be used at the beginning, middle and/or end of your search text: %C57BL%, C57BL% or C57%6%.</li>
			<li>When doing wild card searching for markers or alleles, the wild card can only be used on row 1 of the table.</li>			
		</ul>
	</li>
	<li>You can search by chromosome if you <b>don't</b> tab out of the field.</li>
	<li>The mutant (ES Cell) field <b>cannot</b> be used for searching.</li>
</ul>
Please note that large searches take a long time to clear in browser world; it might be less frustrating to just reload the page. (Remember: items on the clipboard will persist.)

<h4>Searching by Chromosomal Location</h4>
<p>When searching for Chr location without a maker DO NOT tab out before doing the search as this will change the Chr to All. This also holds true when doing a combination search for Chr, State, and/or Compound. </p>

<h4>Searching by Mutant Cell Line</h4>
<p>Note that searching for a mutant cell line in the Mutant 1 or Mutant 2 field is no longer available. To find an annotation to a mutant cell line, the Allele 1 field must have the allele symbol entered in addition to the mutant cell line in Mutant 1. To do this search, enter the allele in Allele 1 and tab out, then enter the mutant cell line in Mutant1 but DO NOT tab out. Hit search. If the mutant 1 field is tabbed out, no results are returned. </p>
<h4>Search Return</h4> 
<ul>
	<li>Only one allele pair per genotype is displayed in the search return.</li>
	<li>For genotypes that contain multiple allele pairs, which allele pair is displayed in the return varies depending on how the genotype was searched for.</li>
</ul>
<h4>Search by Dates</h4>
<ul>
	<li>Valid date formats for searching are:
		<ul>
			<li>mm/dd/yyyy or mm-dd-yyyy</li>
			<li>mm/dd/yy or mm-dd-yy</li>
			<li>yyyy/mm/dd or yyyy-mm-dd</li>
		</ul>
	</li>
	<li>Searching for a range of dates:
		<ul>
			<li>equal to: enter a single date</li>
			<li>greater than: &gt; mm/dd/yyyy</li>
			<li>greater than or equal to: &gt;= mm/dd/yyyy</li>
			<li>less than: &lt; mm/dd/yyyy</li>
			<li>less than or equal to: &lt;= mm/dd/yyyy</li>
			<li>between two dates (inclusive): mm/dd/yyyy..mm/dd/yyyy</li>
			</ul>
	</li>
</ul> 
<h4>Further information</h4>
See <a tabindex="-1" href="http://mgiwiki/mediawiki/index.php/sw:EI_Genotype">Here</a>
 
</div> <!-- helpSectionWrapper -->
