<div id="helpSectionWrapper">

<h2>P-WI Marker Editor Interface functionality</h2>

<h4>General Description</h4>
<p>
The Marker module is a search, display and edit page all in one. A user must be logged in to edit but can search without logging in.

</p><h4>Keyboard Short-cuts</h4>
There are reported issues using these short-cuts with Mac/Firefox. 
<ul>
	<li>Add = Ctrl-Alt-A</li>
	<li>Delete = Ctrl-Alt-D</li>
	<li>Search = Ctrl-Alt-S</li>
	<li>Clear = Ctrl-Alt-C</li>
	<li>Modify = Ctrl-Alt-M</li>
	<li>First = Ctrl-Alt-F</li>
	<li>Previous = Ctrl-Alt-P</li>
	<li>Next = Ctrl-Alt-N</li>
	<li>Last = Ctrl-Alt-L</li>
	
</ul>

<h4>Color-coded Labels</h4>
<ul>
	<li>Thistle = search only
	<li>Green = search & edit
	<li>Blue = read only
	<li>Yellow = current row</li>
</ul>

<h4>Supported Browsers</h4>
<ul>
	<li>Chrome and FireFox are supported on both PC and Mac computers.</li>
	<li>Safari is NOT supported for the PWI. There are features that do not work.</li>
</ul>

<h3>Searching</h3>
<p>
All fields that have labels highlighted in blue/green may be used for searching. They may be searched individually or in 
combination. To include data in a tab in a search: click on the tab to expand it and enter text to be searched. At least one search field must be entered or an error message is displayed. When more than one field is used in a search, it 
is treated as an AND search, e.g. Marker Type = Gene AND Symbol = Pax%.
</p>

<h4>Sort Order</h4>
<ul>
<li>marker type: Gene,DNA Segment,Cytogenetic Marker,QTL,Pseudogene,BAC/YAC end,Other Genome Feature,Complex/Cluster/Region,Transgene </li>
<li>alphanumeric of symbol
</ul> 

<h4>Limit on # of results</h4>
<ul>
<li>The Marker module is currently configured to restrict the number of results to 5000 results. It may be preferable to use more search fields to reduce the number of results returned.</li>
</ul> 

<h4>Wildcards</h4>
<ul>
	<li>Wildcards (%) may be used for most fields. </li>
	<li>A wildcard may be used at the beginning, middle and/or end of your search text: e.g. %paired%, casein% or AK32%43%.</li>
</ul> 

<h4>Search by Accession IDs</h4>
<ul>
	<li>Valid types of IDs: MGI Marker, MGD-MRK IDs, EC IDs and Sequence DB IDs.</li>
	
	
	
</ul>

<h4>Search by Dates</h4>
<ul>
	

	
	
	<li>Valid date formats for searching are:
		<ul>
			<li>mm/dd/yyyy or mm-dd-yyyy</li>
			<li>mm/dd/yy or mm-dd-yy</li>
			<li>yyyy/mm/dd or yyyy-mm-dd</li>
		</ul>
	</li>
	<li>Searching for a range of dates:
		<ul>
			<li>equal to: enter a single date</li>
			<li>greater than: &gt; mm/dd/yyyy</li>
			<li>greater than or equal to: &gt;= mm/dd/yyyy</li>
			<li>less than: &lt; mm/dd/yyyy</li>
			<li>less than or equal to: &lt;= mm/dd/yyyy</li>
			<li>between two dates (inclusive): mm/dd/yyyy..mm/dd/yyyy</li>
			</ul>
	</li>

	

</ul> 

 
<ul>
	<li>The History Event Date field only accepts one format YYYY-MM-DD. The date formats used by other fields do not apply to this field.</li>
	 
</ul>

<h4>Search by Marker History</h4> 
<ul>
	<li>The Created By (User) and Modified By (User) may be searched using the curator (or database load) user name.</li>
	<li>The Creation Date and Modification Date may be searched in any of the following formats:</li> 
	<li>Valid date formats for searching are:
		<ul>
			<li>mm/dd/yyyy or mm-dd-yyyy</li>
			<li>mm/dd/yy or mm-dd-yy</li>
			<li>yyyy/mm/dd or yyyy-mm-dd</li>
		</ul>
	</li>
	<li>A range of dates may also be searched:
		<ul>
			<li>equal to: enter a single date</li>
			<li>greater than: &gt; mm/dd/yyyy</li>
			<li>greater than or equal to: &gt;= mm/dd/yyyy</li>
			<li>less than: &lt; mm/dd/yyyy</li>
			<li>less than or equal to: &lt;= mm/dd/yyyy</li>
			<li>between two dates (inclusive): mm/dd/yyyy..mm/dd/yyyy</li>
		</ul>
	</li>
</ul>

<h4>Search Results</h4>
<p>When a user clicks on the Search button or presses Enter/Return on the keyboard, the results are displayed in a table under the Search button on the right side of the page. The 
		table displays the results in alpha numeric order. The highlighted result is the
		one that is displayed in detail on the left side. A user can highlight a different result 
		by clicking on a different row or using Ctrl+Alt+N or Ctrl+Alt+P to scroll down and up the results. Hovering the cursor in the results table will allow the user to scroll up or down the table of results with a mouse.
	</p>

<h4>Detail Results</h4>

	<p>Detailed information is displayed on the left side of the page in the Marker Detail section.
		By default, the marker highlighted on the right is the first one returned and displayed. (A user must be logged in to Modify a reference.
		When not logged in the Modify, Create and Delete buttons are greyed out. Create stays greyed out unless new data is being entered.)</p>
		
<h3>Adding New Data</h3>
<ul>
        <li>Adding a new marker requires a unique symbol, a name and a chromosome. If no marker type and marker status are chosen, the default marker type is gene and the default marker status is official.
	<li>J:23000 is the default reference if you don't enter one when creating a new marker.</li>
<li>Feature Type (TDC) can be selected from the feature type drop down list. The marker types and which feature types are allowed with them are listed at the bottom of this document.</li>
	<li>A marker cannot be created if the symbol already exists.</li>
	<li>Adding a marker of type "Gene" and status "Official" always creates a wild-type allele for the gene.</li>
	<li>If Chromosome is set to "UN" the cmOffset is set to -999, if Chromosome is not set to "UN" the cmOffset is set to -1.</li>
</ul>

<h3>Modifying fields</h3>
<ul>
	<li>Marker Status cannot be changed to 'Withdrawn' by the Modify button.</li>
	
	<li>Marker Status cannot be changed from "Official" to "Reserved" if the marker has annotations associated to it.</li>
	<li>If Marker Status is being changed from "Reserved" to "Official" and Marker Type is "Gene" and a Wild-type allele does not exist and Marker 
	Name does not contain (mt, withdrawn, dna sequence, EST, expressed sequence, cDNA sequence, gene model, hypothetical protein, 
	ecotropic viral intergration site, or viral polymerase), then create the wild-type allele.</li>
	<li>The Modify button is used to update/create/remove editor notes, sequence notes, marker revision notes, strain-specific notes, location notes, change the chromosome, marker type, marker status, add/remove feature types and correct typos in the symbol and name.</li>
</ul>

<h3>Sections of the page</h3>
<h4>Right side</h4>
<ul>
    <li>This part of the page contains the searchable fields MGI ID, Created by, Modified by, Creation Date, Modification Date and the action buttons Search, Clear, Reset Search, Create, Modify, Delete.</li><li>Clear resets the page. Reset Search removes items returned from a search but keeps the last search parameters. You must be logged in to use Create, Modify and Delete.</li>
</ul>
	<h4>Marker Detail</h4>
<ul>
    <il>The top left part of the page contains drop down lists of Marker Type, Marker Status, Chromosome and Feature Type. Below the drop down lists are Symbol, Name, Cytogenetic Band and cM Position fields.
</il></ul>
<h4>Notes Tabs</h4><ul><li>Editor Notes, Sequence Notes, Marker Revision Notes, Strain-Specific Notes, and Location Notes are all tabs immediately under the upper left Marker Detail section.</li><li>Clicking on a note will open a text field and allow creation of a new note. Click the 
    <b>Modify</b>
    button to save any changes.</li><li>If a note is already present, the tab title is in green instead of black font.</li><li>The Strain-Specific Notes appear on the public marker detail pages.</li></ul>
<ul></ul>
<ul></ul>

<h4>History Table</h4>
<ul>
<li>This table appears below the notes tabs. It contains the fields #, Symbol, Name, Date, J#, Citation, Event, Reason, Modified By. The fields populate when a Utility function (see below) is used.</li>
<li>Even though the Symbol, Name, Date, J#, Event, and Reason fields are all editable, they should not be under normal circumstances and in accordance with best practices.</li>
<li>The Symbol and J# fields are validated. Validation happens once you tab out of the field or click in another field. If the field does not validate the user is presented with an error. The Modify button is disabled upon a validation error and re-enabled once a validated entry is entered. As a visual clue, the Modify button is greyed out when there is a validation error.</li>
<li>The Citation and Modified By fields are not editable.</li>
<li>The # field can be edited but must validate before it can be modified: Numbers can be switched among rows, but no two rows are allowed to have the same number and a new number larger than that of the last row cannot be used. Once modified, the rows are resorted by # numbers into their new sort order.</li>
<li>To add a new row, click the <b>Add</b> button, then click <b>Modify</b>.</li>
<li>To delete an existing row, click the <b>X</b></b> button, then click <b>Modify</b>.</li>
</ul><h4>Other Tabs</h4>

<h5>Synonyms Tab</h5>
<ul>
<li>All input fields are searchable.</li>
<li>Results are sorted by Synonym Type (in same order as pull down menu). Secondary sort is by Synonym alpha-numerically.</li>
<li>J# field is validated.</li>
<li>To add a new row, click the <b>Add</b> button, then click <b>Modify</b>.</li>
<li>To delete an existing row, click the <b>X</b></b> button, then click <b>Modify</b>.</li>
</ul>
<h5>References Tab</h5>
<ul>
<li>All input fields are searchable. A wildcard is allowed for the Citation field.</li>
<li>Results are sorted by Type (General first). Secondary sort is by J#.</li>
<li>References can be added/deleted but not modified.</li>
<li>J# field is validated.</li>
<li>To add a new row, click the <b>Add</b> button, then click <b>Modify</b>.</li>
<li>To delete an existing row, click the <b>X</b></b> button, then click <b>Modify</b>.</li>
</ul>
<h5>AccIDs (Editable) Tab</h5>
<ul>
<li>All input fields are searchable.</li>
<li>Results are sorted by Acc Name (alpha-numerical order). Secondary sort is AccID.</li>
<li>J# field is validated.</li>
<li>To add a new row, click the <b>Add</b> button, then click <b>Modify</b>.</li>
<li>To delete an existing row, click the <b>X</b></b> button, then click <b>Modify</b>.</li>
</ul>
<h5>AccIDs (Other) Tab</h5>
<ul>
<li>This tab is display only.</li>
<li>Results are sorted by Acc Name (alpha-numerical order). Secondary sort is AccID.</li>
<li>You can use your browser's find feature to locate a particular ID or any other field in this section.</li>
</ul>
<h5>MGI IDs Tab</h5>
<ul> 
<li>This tab is display only.</li><li>The current MGI ID is the first one listed.</li>
<li>Other IDs in this section are MGD-MRK IDs and old MGI IDs.</li>
</ul>
<h5>TSS Tab</h5>
<ul>
<li>This tab is display only.</li>
<li>Results are sorted by ID (alpha).</li>
<li>You can use your browser's find feature to locate a particular ID or any other field in this section.</li>
</ul>
<h5>STS markers Tab</h5>
<ul>
<li>This tab is display only.</li>
<li>Results are sorted alphabetically.</li>
<li>You can use your browser's find feature to locate a particular ID or any other field in this section.</li>
</ul>
<h3>Utilities</h3>
<h4>Rename</h4>
<ul>
<li>Select Rename from the event list.</li>
<li>Select an Event Reason from that list.</li>
<li>Enter a valid J#.</li>
<li>Tab to the New Symbol field. Pick a unique symbol to be sure it will validate.</li>
<li>Enter a symbol name.</li>
<li>You can uncheck the Add marker as synonym button if the symbol to be renamed is a gene model or otherwise uninformative.</li>
<li>Click <b>Process</b>. The name and symbol are updated in the top section. History rows are added.</li>
</ul>
<h4>Merge</h4>
<ul>
<li>Select Merge from the event list.</li>
<li>Select an Event Reason or leave not specified.</li>
<li>Enter a valid J#.</li>
<li>Tab to the Symbol field. Enter a valid symbol. The AccID field will populate. If the marker to be merged into is on a different chromosome, a popup alert will appear.</li>
<li>Leave checked or uncheck the Add marker as synonym button.</li>
<li>Click <b>Process</b>. The name field will update to 'withdrawn, = new symbol.' The history rows will disappear from this marker. A new history row will be created in the merged into marker.</li>
</ul>

<h5>Merging Rules</h5>	
<ul>
<li>You cannot Merge a Symbol into itself.</li>
<li>You cannot Merge if both Symbols contain a Marker Detail Clip.</li>
<li>If new/merge Symbol has a chromosome = UN, then database will try to update the new symbol with the old symbols chromosome.</li>
<li>The old symbol's Cytogenetic Band and cM Offset are used to update Band/Offset of new symbol.</li>
<li>If the old symbol has a wild type allele, then that is merged into the new symbol if it doesn't have one.</li>
<li>If the old symbol is Rik, then synonym type = similar, else exact.</li>
</ul>
<h4>Allele Of</h4>
<ul>
<li>Select Allele of from the event list.</li>
<li>Select an event reason or leave not specified.</li>
<li>Enter a valid J#.</li>
<li>Tab to the Symbol field. Enter a valid symbol. The AccID field will populate once you Tab again or use the mouse to click on another field.</li>
<li>Leave checked or uncheck the add marker as synonym button.</li>
<li>Click <b>Process</b>. The name field will update to 'withdrawn, = new symbol.' The history rows will disappear from this marker. A new history row will be created in the parent marker.</li>
</ul>
<h4>Delete</h4>
<ul>
<li>Select Delete from the event list.</li>
<li>Enter a valid J#.</li>
<li>Click <b>Process</b>. The name in the top section is replaced with 'withdrawn.' A history row is created.</li>
</ul>

<h4>Feature Type hierarchy</h4>
<ul style="list-style-type: square;">	
<li>Gene--</li></ul>
<ul style="list-style-type: circle;">
<li>gene</li>
<li>gene segment</li>
<li>heritable phenotypic marker</li>
<li>lncRNA gene</li>
<li>antisense lncRNA gene</li>
<li>sense intronic lncRNA gene</li>
<li>sense overlapping lncRNA gene</li>
<li>bidirectional promoter lncRNA gene</li>
<li>lincRNA gene</li>
<li>miRNA gene</li>
<li>non-coding RNA gene</li>
<li>protein coding gene</li>
<li>rRNA gene</li>
<li>ribozyme gene</li>
<li>RNase P RNA gene</li>
<li>RNase MRP RNA gene</li>
<li>tRNA gene</li>
<li>scRNA gene</li>
<li>snRNA gene</li>
<li>snoRNA gene</li>
<li>SRP RNA gene</li>
<li>telomerase RNA gene</li>
<li>unclassified gene</li>
<li>unclassified non-coding RNA gene</li>
</ul>
<ul style="list-style-type: square;">	
    <li>DNA Segment--no feature type</li></ul>	
	
<ul style="list-style-type: square;">	<li>Cytogenetic Marker--</li></ul>
<ul style="list-style-type: circle;">
<li>chromosomal deletion</li>
<li>chromosomal deletion</li>
<li>chromosomal duplication</li>
<li>chromosomal inversion</li>
<li>chromosomal translocation</li>
<li>chromosomal transposition</li>
<li>insertion</li>
<li>reciprocal chromosomal translocation</li>
<li>Robertsonian fusion</li>
<li>unclassified cytogenetic marker</li>
</ul>
<ul style="list-style-type: square;"><li>QTL--no feature type</li></ul>	
<ul style="list-style-type: square;"><li>Pseudogene--</li></ul>
<ul style="list-style-type: circle;">
<li>polymorphic pseudogene</li>
<li>pseudogene</li>
<li>pseudogenic gene segment</li>
<li>pseudogenic region</li>
</ul>
<ul style="list-style-type: square;"><li>BAC/YAC end--no feature type</li></ul>
	

<ul style="list-style-type: square;"><li>Other Genome Feature--</li></ul>
<ul style="list-style-type: circle;">	
<li>CpG island</li>
<li>endogenous retroviral region</li>
<li>minisatellite</li>
<li>mutation defined region</li>
<li>other genome feature</li>
<li>promoter</li>
<li>retrotransposon</li>
<li>telomere</li>
<li>TSS cluster</li>
<li>unclassified other genome feature</li>
<li>origin of replication</li>
<li>transcriptional cis regulatory region</li>
</ul>
<ul style="list-style-type: square;"><li>Complex/Cluster/Region--no feature type</li></ul>
<ul style="list-style-type: square;"><li>Transgene--no feature type</li></ul><br>	
If new feature types are created, the code may need to be modified to add the types to the hierarchy. Marker types are only compatible with feature types that are subsets of their type. As noted above, some marker types do not get feature (TDC) types, such as QTL, Complex/Cluster/Region, Transgene, BAC/YAC end, and DNA segment.	

</div>
